ADBDEV-2160 Fix CTE with DML operations on replicated tables

bimboterminator1 opened the pull request:
Add support of CTE with modifying DML operations on replicated tables. 

Several issues could occur while running queries with writable CTEs, which
modify replicated tables. This patch proposes solution for main aspects of
handling the modifying CTE over replicated tables. Those issues appeared due to
inability of major interfaces to handle locus type CdbLocusType_Replicated and
flow type FLOW_REPLICATED, which is set for the result of modifying operation
over replicated table inside the adjust_modifytable_flow function. 

### 1. Plain SELECT from CTE with locus Replicated

SELECT statement from modifying CTE over replicated table with RETURNING clause
didn't gather the CTE's result, i.e planner didn't add the focusing motion node
at the top of the CTE's subplan. This happend because apply_motion function
didn't expect that the result of the plan with flow->flotype = FLOW_REPLICATED
could be returned to frontend. The absence of the top motion node leads to an
empty final result set, no matter how many rows were processed by
INSERT/UPDATE/DELETE beneath.

This patch lets the apply_motion function add the Explicit Gather Motion to
the top of writable CTE plan by extending the if-clause, which decides whether
the result need to be brought to dispatcher, with the check
plan->flow->flotype is FLOW_REPLICATED.

### 2. Choosing Join locus type

The planner failed to form a plan when trying to choose
correct final locus type for join operation between modifying CTE over
replicated table and any other relation. The planner were failing either with
errors, because cdbpath_motion_for_join function could not choose a proper
join locus, or produced invalid suboptimal plan when handling Replicated locus.
Therefore, this patch reconsiders the join logic with locus Replicated.
The main principle to decide join locus of Replicated locus with others is: the
slice which does the modifying operation on replicated table must be executed on
all segments with data. And we also have to keep in mind that Redistributing or
Broadcasting Replicated locus will lead to data duplication. The joins are
performed in the following way:
###  SegmentGeneral + Replicated

When the join between CdbLocusType_Replicated and CdbLocusType_SegmentGeneral
occured, the planner failed with the assert, which checks that the common
UPDATE/DELETE ... FROM/USING (modifying operation with join) of a replicated
table takes place (root->upd_del_replicated_table == 0 condition in the
cdbpath_motion_for_join function).This assert was added by the author without
proper test coverage, and, currently, there is no case when this check can be
true.

Without the assert the planner produced the valid plan, however it could cut the
number of segments in the final locus. In this context locus Replicated was not
correctly handled, therefore one need to add the proper logic which takes into
account the writable CTE case.

This patch allows joins between CdbLocusType_Replicated and
CdbLocusType_SegmentGeneral in the following way: for the cases, when
SegmentGeneral's number of segments is greater or equal than Replicated's, the
final join locus type becomes Replicated as well. Otherwise we gather both
Replicated and SegmentGeneral parts to SingleQE in order to perform join at
single segment.
### Replicated + General

The logic related to General locus remained unchanged. The join locus type
becomes Replicated.
### Replicated + SingleQE

If the join between CdbLocusType_Replicated and CdbLocusType_Entry or
CdbLocusType_SingleQE takes place, the Replicated locus will be brought to Entry
or to SingleQE respectively.
### Replicated + Partitioned

If the join between CdbLocusType_Replicated and CdbLocusType_Hashed or
CdbLocusType_Strewn takes place, and the number of segments for Replicated locus
is equal to number of segments for other locus, and it's not an outer join, the
join will be performed with Hashed or Strewn locus respectively. Otherwise, 
both parts are brought to SingleQE.
### Replicated + Replicated

Join between 2 parts with Replicated locus leads to final join locus being also
Replicated.

### 3. Creating motion from path or plan with Replicated locus

When locus type is changed inside cdbpath_motion_for_join
(Replicated-->SingleQE),  another errors occur inside the
cdbpath_create_motion_path function. This function did not allow motion from
replicated to singleton locus. This patch extends the condition  that validates
source locus type, and allows the movement from Replicated locus to singleton.

When we try apply a motion to the plan with plan->flow->flotype value
FLOW_REPLICATED via make_motion_gather function, the planner will fail
with the assert, because such flow is unexpected in this function. For example,
the assertion error was thrown for the case when  ORDER BY  was applied to the
result of modifying CTE over replicated tables.  When it was called there was
the assert checking that plan->flow->flotype was either FLOW_PARTITIONED or
FLOW_SINGLETON. And, therefore, it failed for FLOW_REPLICATED. Without asserts
the Explicit Gather Motion is applied correctly. This patch changes this assert
by allowing function to pass with FLOW_REPLICATED.

### 4. UNION ALL which includes Replicated path

Here two issues occured. When UNION ALL contained a path with locus Replicated,
and other path, which is propagated to less number of segments than Replicated
is, the UNION target locus is aligned to the smallest number of segments among
the operands inside the set_append_path_locus function, what leads to segments
reduction of locus Replicated. This behaviour is invalid, because Replicated
locus must be executed on all segments with data. Therefore this patch does
not allow to make alignment of final UNION locus in case when target locus
is Replicated.

When UNION ALL were executed with operands, one of which had locus
Replicated and the other had locus SegmentGeneral, the planner were failing
with assertion error from cdbpath_create_motion_path. Final locus of UNION
was decied to be Replicated, and therefore, all other arguments had to become
Replicated as well. The issue was that  cdbpath_create_motion_path function
forbidden operations of moving SegmentGeneral path to Replicated locus.

This patch solves the issue by making SegmentGeneral locus stay unchanged
in case when SegmentGeneral's number of segment is greater or equal
than Replicated's. In this case no motion needed. Otherwise SegmentGeneral path
is broadcasted from single segment.

### 5. Volatile functions handling

When a query had quals, which were applied to the modifying CTE and
contained volatile functions, the final plan became invalid, because its
execution led to volatile tuple set on different segments (however, we expect
the set of Replicated locus be the same everywhere). That could produce
inconsistent results in case of joins, SubPlans. And the issue is not only with
quals, but it can occur for volatile targetlists or joinquals.

This patch solves this issue by total prohibition of volatile functions applied
to the plan or path with locus Replicated. The funcions
turn_volatile_seggen_to_singleqe and subquery_planner were extended by the
condition, which cheks whether given locus type is Replicated and there are
volatile functions over it. If condition is satisfied the proper error is thrown

The changes cover volatile targetlist, returning list, plan quals or join quals,
having clauses.

### 6. Replicated locus with different number of segments inside SubPlans

The last issue solved by this patch occured when the modifying CTE was
referenced inside any SubPlan. In this case ParallelizeSubplan and
ParallelizeCorrelatedSubPlanMutator functions tried to broadcast already
replicated plan, what could lead to data dublication. Therefore, one need to
prevent these funcions from broadcasting the result with Replicated locus.

This patch modifies both functions by adding the condition, depending on which
the planning goes on or the error is thrown. If plan's locus type is
CdbLocusType_Replicated and its numsegments is equal to number of segments of
target distribution, the broadcast doesn't occur. If the number of segments is
different, the error is thrown.


Co-Authored-By: Alexey Gordeev <goa@arenadata.io>

InnerLife0 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -368,7 +376,12 @@ cdbpath_create_motion_path(PlannerInfo *root,
 		else if (CdbPathLocus_IsReplicated(locus))
 		{
-			/* Assume that this case only can be generated in DML statement. */
+			/* Motion is allowed in two cases. Firstly, when replicated table
```
Wrong multiline comment style. Please check all patch changes with pgindent.

bimboterminator1 replied:
Done

InnerLife0 commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -427,304 +427,412 @@ DROP TABLE r;
+drop table if exists t_repl;
+drop table if exists t_part;
+--end_ignore
+create table t_repl (i int, j int) distributed replicated;
```
What was the reason of renaming? The commit diffs are meaningless huge with this changes.

bimboterminator1 replied:
The main reason is to make namings simple and readble. The "with_dml_dr", with_dml" names, in my opinion, don't reflect table disitribution type.

InnerLife0 commented:
Talking about first commit, I approve it, but I should say it's much better to separate comments refactoring to another commit. Your commit contains changes which are not related to code fix. For reviewer, it may be hard to isolate just subset of changes he needs to review manually.

I have left a few comments for the second commit and will continue with review of patch logic next time.

RekGRpth commented:
> As for multiple references to modifying CTEs, they can lead to duplicated DML operations during execution. Propose chancges add to the `set_cte_pathlist` function the if clause, which checks if CTE contain modifying operations and `cte->cterefcount > 1`. If this condition is true, the elog with error message "too much refs to non-SELECT CTE" is called.

Have you considered using the `gp_cte_sharing` flag in this case?

InnerLife0 commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -544,17 +549,91 @@ with cte as (
     insert into with_dml_dr select i, i * 100 from generate_series(1,5) i
     returning i
 ) select count(*) from cte left join (select * from with_dml limit 5) lmt on cte.i = lmt.i;
+explain (costs off)
```
We probably miss the same query, but without EXPLAIN.

bimboterminator1 replied:
Added the test

InnerLife0 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1287,24 +1298,86 @@ cdbpath_motion_for_join(PlannerInfo *root,
+				/* Otherwise gather both of them to a SingleQE */
+				else
+				{
+					int		numsegments = CdbPathLocus_NumSegments(replicated->locus);
```
Do we need CommonSegments() call or some assertion here?

bimboterminator1 replied:
We get to that scope if Replicated numsegments < Partitioned numsegments. CommonSegments will return the min(replicated->locus,other->locus) = replicated->locus.numsegments. 
Assert is also redundant for this case

InnerLife0 commented on src/test/regress/expected/with_clause.out:
```diff
@@ -2597,29 +2603,29 @@ with cte as (
 ) select count(*) from cte left join (select * from with_dml limit 5) lmt on cte.i = lmt.i;
-                                                      QUERY PLAN                                                      
-----------------------------------------------------------------------------------------------------------------------
- Explicit Gather Motion 3:1  (slice3; segments: 3)  (cost=39.34..39.35 rows=1 width=8)
```
Am I wrong, or this plan became less optimal?

bimboterminator1 replied:
We avoid broadcast in this case

InnerLife0 commented on src/test/regress/expected/with_clause.out:
```diff
@@ -2716,11 +2918,98 @@ with cte as (
+) select count(*) from cte left join t_strewn on cte.i = t_strewn.i;
+                             QUERY PLAN                              
+---------------------------------------------------------------------
+ Aggregate
```
Curious if it's possible to calc Aggregate on segments.

bimboterminator1 replied:
The behaviour is the same as for SegmentGeneral case. 
**SegmentGeneral case**
```
explain (costs off) select count(*) from with_dml_dr_seg1 left join t_strewn using(i);
                 QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Hash Right Join
         Hash Cond: (t_strewn.i = with_dml_dr_seg1.i)
         ->  Gather Motion 3:1  (slice1; segments: 3)
               ->  Seq Scan on t_strewn
         ->  Hash
               ->  Gather Motion 1:1  (slice2; segments: 1)
                     ->  Seq Scan on with_dml_dr_seg1
 Optimizer: Postgres query optimizer
(9 rows)

```
Both relations are brought to SingleQE, and this behaviour becomes suboptimal if target segment is QD on master. Better approach is to gather other relation to segments where the Replicated locus is. 
Using `CdbPathLocus_MakeSingleQE`  for that purpose is not reliable enough and doesn't always work, as we don't really decide the final segment the node will be executed at. 
```
else
{
int		numsegments = CdbPathLocus_NumSegments(replicated->locus);
//CdbPathLocus_MakeSingleQE(&replicated->move_to, numsegments);
CdbPathLocus_MakeSingleQE(&other->move_to, numsegments);
}


explain (costs off)  with cte as (                                           
    insert into with_dml_dr_seg1 select i, i * 100 from generate_series(1,5) i
    returning i               
) select count(*) from cte join t_strewn on cte.i = t_strewn.i;
                         QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Hash Join
         Hash Cond: (t_strewn.i = with_dml_dr_seg1.i)
         ->  Gather Motion 3:1  (slice1; segments: 3)
               ->  Seq Scan on t_strewn
         ->  Hash
               ->  Insert on with_dml_dr_seg1
                     ->  Function Scan on generate_series i
 Optimizer: Postgres query optimizer
(9 rows)
```  
Or we can broadcast other relation. This thing is likely to be executed on segment hosts. It is good, but it is broadcast.
```
else
{
int		numsegments = CdbPathLocus_NumSegments(replicated->locus);
//CdbPathLocus_MakeSingleQE(&replicated->move_to, numsegments);
//CdbPathLocus_MakeSingleQE(&other->move_to, numsegments);
CdbPathLocus_MakeReplicated(&other->move_to, numsegments);
}
explain (costs off)  with cte as (                                                
    insert into with_dml_dr_seg1 select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join t_strewn on cte.i = t_strewn.i;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Explicit Gather Motion 1:1  (slice2; segments: 1)
   ->  Aggregate
         ->  Hash Join
               Hash Cond: (with_dml_dr_seg1.i = t_strewn.i)
               ->  Insert on with_dml_dr_seg1
                     ->  Function Scan on generate_series i
               ->  Hash
                     ->  Broadcast Motion 3:1  (slice1; segments: 3)
                           ->  Seq Scan on t_strewn
 Optimizer: Postgres query optimizer
(10 rows)

```

Well, here the choice between SingleQE or Broadcast operation. If our relation with_dml_dr_seg1 was distributed on 2 semgents and t_strewn on 3, there would be Gather Motion 3:1  and Explicit Gather Motion 2:1 to master (probably) for SingleQE case. And if we chose broadcast, there would be only broadcast motion 3:2, but on segments.

bimboterminator1 replied:
The example for SegmentGeneral is the same if processing common (not outer) join.
```
explain (costs off) select count(*) from with_dml_dr_seg1 join t_strewn using(i);
                 QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Hash Join
         Hash Cond: (t_strewn.i = with_dml_dr_seg1.i)
         ->  Gather Motion 3:1  (slice1; segments: 3)
               ->  Seq Scan on t_strewn
         ->  Hash
               ->  Gather Motion 1:1  (slice2; segments: 1)
                     ->  Seq Scan on with_dml_dr_seg1
 Optimizer: Postgres query optimizer
(9 rows)
```

bimboterminator1 commented:
Query for local join.
```
else if (CdbPathLocus_NumSegments(replicated->locus) >=
	    CdbPathLocus_NumSegments(other->locus))
	return other->locus;
```

```
create table t_repl (i int, j int) distributed replicated;
create table t_hashed (i int, j int) distributed by (i);
set allow_system_table_mods = on;
update gp_distribution_policy set numsegments = 2 where localoid = 't_hashed'::regclass::oid;
explain (costs off)  with cte as (                                                
    insert into t_repl select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join t_hashed on cte.i = t_hashed.i;

```

Stolb27 commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -617,7 +617,8 @@ apply_motion(PlannerInfo *root, Plan *plan, Query *query)
 				if ((plan->flow->flotype == FLOW_PARTITIONED ||
 					(plan->flow->flotype == FLOW_SINGLETON &&
-					 plan->flow->locustype == CdbLocusType_SegmentGeneral)) &&
+					 plan->flow->locustype == CdbLocusType_SegmentGeneral) ||
+					 query->hasModifyingCTE) &&
```
Initially, this condition said: if data on segments (partitioned or replicated flow) and we want it on coordinator (non-parallel cursor case), append gather motion. Why existing of Modifying CTE somewhere in plan leads to requirement to gather data unconditionally?

InnerLife0 replied:
Quick test without this condition.
[regression.diffs.txt](https://github.com/arenadata/gpdb/files/11755218/regression.diffs.txt)

The example query.
```
explain (costs off)
with cte as (
    update with_dml_dr set j = j + 1 where i <= 5 returning j
) select count(*) from cte;
             QUERY PLAN              
-------------------------------------
 Aggregate
   ->  Update on with_dml_dr
         ->  Seq Scan on with_dml_dr
               Filter: (i <= 5)
 Optimizer: Postgres query optimizer
(5 rows)

with cte as (
    update with_dml_dr set j = j + 1 where i <= 5 returning j
) select count(*) from cte;
 count 
-------
(0 rows)
```
For now, I can't remember all the reasons of why I decided to add this check and made it such wide.

bimboterminator1 replied:
The `apply_motion` function is called in order to apply motion node to the top-level plan tree if necessary. For child plan nodes this function is not called.
As for modifying CTE's cases, `query->hasModifyingCTE` flag is set only for top-level select statement. And in the patch the `bringResultToDispatcher = true` will take place always when such CTE is present. However there is the finit set of possible Flow types at all: `FLOW_SINGLETON`, `FLOW_REPLICATED`, `FLOW_PARTITIONED`. 
For `FLOW_PARTITIONED` the condition will be true and it's ok independently from CTE presence.
For `FLOW_SINGLETON` there is the following logic:
1. **With our patch**:
`bringResultToDispatcher = true` because of `query->hasModifyingCTE` flag and `focusPlan` is [called](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L652-L680). However [inside](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbllize.c#L1116-L1118) nothing changes. Then we [get](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L682)  to `apply_motion_mutator` function, where for most queries with modifying CTE (for examples in tests section of commit) this part of code is [executed](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L964-L969). Therefore, checking `query->hasModifyingCTE` in if-clause is really **unneseccary**  as it does not affect anything in this case.
2. **Without our patch** we get to `apply_motion_mutator` without `focusPlan`, and have the same [situation](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbmutate.c#L964-L969).


And `FLOW_REPLICATED` can happen in top-level plan node only for our modifying CTE case (unproven hypothesis), and here we should add motion plan if `query->hasModifyingCTE` is true.

Therefore, `query->hasModifyingCTE` doesn't really break general behavior. And to make it a little bit clearer I can change the if-clause the following way:
```
if ((plan->flow->flotype == FLOW_PARTITIONED ||
    (plan->flow->flotype == FLOW_SINGLETON &&
     plan->flow->locustype == CdbLocusType_SegmentGeneral)) ||
    (plan->flow->flotype == FLOW_REPLICATED &&    <------ add check here
     query->hasModifyingCTE)) && 
    !root->glob->is_parallel_cursor)
          bringResultToDispatcher = true;

```
I have already run the CI tests for version with such if-clause, tests have passed  without problems. So, should I add that?

Stolb27 replied:
What about handling any cases of FLOW_REPLICATED because we know that they handled properly (by adding Explicit Gather Motion)? Do we really need this obscure condition (`query->hasModifyingCTE`)?

bimboterminator1 replied:
I assume, that main questions are: Is the gather motion always explicit when the sender's locus type is Replicated? Will it be safe to remove such kind of restrictions on replicated locus? 
Well, consider the executor's code. [Here](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/executor/nodeMotion.c#L1056-L1067) it is said, that gather motion is marked as explicit for the locus Replicated. The concept of Explicit Gather Motion was added in this [commit](https://github.com/arenadata/gpdb/commit/fb7247b93bc43c534ae2adbaedb45d8947adefb2),which extended the abilities of locus Replicated, and this was done for RETURNING clause support in modifying queries. I guess, that all restrictions about replicated locus had been made before adding Explicit Gather Motion functionality, and the restrictions were weakened only for DML over replicated tables. And, imho, the other cases, where we need explicit data gathering, simply do not exists or irrelevant. (except modifying CTEs, of course).  I think, we can allow explicit gathering everywhere it is needed, because it's now safe.  Therefore, we can just add FLOW_RAPLICATED here without consequences.

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -225,17 +241,25 @@ cdbpath_create_motion_path(PlannerInfo *root,
 
-		/* replicated-->singleton would give redundant copies of the rows. */
+		/*
+		 * replicated-->singleton would give redundant copies of the rows.
+		 * But there is an exception - DML over replicated table should pass
+		 * the check and add motion.
```
The similar question. We know that sometimes locus Replicated may be successfully moved to Bottleneck (by focusing with Gather Motion). What sort of plans we want to filter here if they will be handled properly?

bimboterminator1 replied:
This restriction appeared before the Explicit Gather Motion logic. For now, I assume, it is safe to do that.
I have removed this restriction for replicated locus, and also have added only the FLOW_REPLICATED to the if-clause (in apply_motion code). All tests passed #528

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1109,19 +1140,30 @@ cdbpath_motion_for_join(PlannerInfo *root,
 				 */
-				Assert(CdbPathLocus_NumSegments(segGeneral->locus) <=
-					   CdbPathLocus_NumSegments(other->locus));
+				Assert((root->upd_del_replicated_table > 0 &&
+					   CdbPathLocus_NumSegments(segGeneral->locus) <=
+					   CdbPathLocus_NumSegments(other->locus)) ||
+					   hasModifyingCTESelectRecur(root));
```
Can you explain difference in handling UPDATE/DELETE with and without wrapping CTE? Code before patch assumes that it's possible to face join between Replicated and SegmentGeneral distribution. And the only expected case is number of segments of SegmentGeneral is less or equal of number of segments of Replicated. Why do we ready to ignore this strange limitation inside CTE?

bimboterminator1 replied:
This limitation was originally aimed at handling the case of common UPDATE/DELETE query. When doing `UPDATE ... FROM` and updating specific rows on join condition, one need to be sure that DML operation will perform on all the segments with data. This `<=` constraint was introduced in this [commit](https://github.com/greenplum-db/gpdb/commit/b32d6110d9a9642c9f675525b7312e3b927b377b) and is used for the situation when replicated table with SegmentGeneral locus is updated on join condition with other relation (locus Replicated). And such restriction means that one cannot update SegmentGeneral table by join with table, which is distiributed on less number of segments than the table one want to update. 
For our situation with modifying CTE this limitaion is unnecessary, because we just want to perform common join with 2 relations: replicated return of CTE (locus Replicated) and replciated table (SegmentGeneral). And here it is appropriate to `CdbPathLocus_NumSegments(segGeneral->locus)` be `>`  `CdbPathLocus_NumSegments(other->locus)`. And moreover the situation when we have simultaniously 2 DML statments (`with cte as (update ... returning) update t_repl set ... from cte`) is currently not allowed.

Stolb27 replied:
> This <= constraint was introduced in this [commit](https://github.com/greenplum-db/gpdb/commit/b32d6110d9a9642c9f675525b7312e3b927b377b) and is used for the situation when replicated table with SegmentGeneral locus is updated on join condition with other relation (locus Replicated). ... And moreover the situation when we have simultaniously 2 DML statments (with cte as (update ... returning) update t_repl set ... from cte) is currently not allowed.

I think it's the only reason to leave this condition as assertion. It's impossible now, but in case of implementation multi-DML queries your modification make this asserion useless. It's always be true in case of presesnce any modifing DML inside CTE. I think we should separate case when target relation of DML operation has Segment General distribution to validate with asserion and common join between Segment General and Replicated when validation is uselss. Comment for the 1st case should be actualized too.

Stolb27 replied:
Comments and modification strict below contradicts with your statement too. It's useless and potentially danger to modify numsegments for replicated distribution generated by DML operation.

bimboterminator1 replied:
Have separated those 2 cases

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1237,25 +1279,96 @@ cdbpath_motion_for_join(PlannerInfo *root,
+		else if (hasModifyingCTESelectRecur(root))
+		{
+			/* Entry has to be broadcasted */
+			if (CdbPathLocus_IsEntry(other->locus))
+				CdbPathLocus_MakeReplicated(&other->move_to,
```
The next strange (join on both segments but throw away on one of them) plans are originated by this decision. Can you explain it?

```
i2160=# explain (costs off) with cte1 as (update d2 set c1 = 42 returning *) select * from cte1 join gp_segment_configuration on dbid = c1;
                         QUERY PLAN
------------------------------------------------------------
 Explicit Gather Motion 2:1  (slice2; segments: 2)
   ->  Hash Join
         Hash Cond: (gp_segment_configuration.dbid = d2.c1)
         ->  Broadcast Motion 1:2  (slice1)
               ->  Seq Scan on gp_segment_configuration
         ->  Hash
               ->  Update on d2
                     ->  Seq Scan on d2
```

bimboterminator1 replied:
Yep, I share your point of view. I can suggest 2 alternatives:
1.  
``` 			
if (CdbPathLocus_IsEntry(other->locus))
CdbPathLocus_MakeReplicated(&other->move_to, 1); 
```
```
explain (costs off) with cte as (update t_repl set i = 2 returning i) select * from cte join gp_segment_configuration on dbid = cte.i;
                          QUERY PLAN                           
---------------------------------------------------------------
 Explicit Gather Motion 1:1  (slice2; segments: 1)
   ->  Hash Join
         Hash Cond: (t_repl.i = gp_segment_configuration.dbid)
         ->  Update on t_repl
               ->  Seq Scan on t_repl
         ->  Hash
               ->  Broadcast Motion 1:1  (slice1)
                     ->  Seq Scan on gp_segment_configuration
 Optimizer: Postgres query optimizer
(9 rows)
```
2.  Make the [same](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbpath.c#L1163-L1172) way as for SegmentGeneral + Entry
```
explain (costs off) with cte as (update t_repl set i = 2 returning i) select * from cte join gp_segment_configuration on dbid = cte.i;
                       QUERY PLAN                        
---------------------------------------------------------
 Hash Join
   Hash Cond: (t_repl.i = gp_segment_configuration.dbid)
   ->  Explicit Gather Motion 3:1  (slice1; segments: 3)
         ->  Update on t_repl
               ->  Seq Scan on t_repl
   ->  Hash
         ->  Seq Scan on gp_segment_configuration
 Optimizer: Postgres query optimizer
(8 rows)

```

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1237,25 +1279,96 @@ cdbpath_motion_for_join(PlannerInfo *root,
+				 * bring locus Replicated to singleQE in order to perform join
+				 * on single segment
+				 */
+				int		numsegments = CdbPathLocus_CommonSegments(replicated->locus,
```
Why not CdbPathLocus_MakeSingleQE to be easy?

bimboterminator1 replied:
I blindly stole logic from SegmentGeneral [case]([same](https://github.com/arenadata/gpdb/blob/6ed03a1d78ef8ab6560dc9a5b40a47cd75faaa75/src/backend/cdb/cdbpath.c#L1163-L1172)). There was an explicit locus assignment because the case included both Entry and SingleQE cases. And in my patch I separated these cases (Now, I consider it to be wrong).

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1109,19 +1140,30 @@ cdbpath_motion_for_join(PlannerInfo *root,
+				 * its all, and SELECT part of query be executed on less
+				 * numsegments. "Explicit Gather Motion" will use less
+				 * numsegments in execMotionSender(), so our result will not
+				 * be flacky because of new segments without a data.
```
What did you mean by this part of the comment? Explicit Gather Motion may be used only to focus tuples to the single segment, but earlier you told about replicated table distributed on less numsegments (no obligatory on the single segment).

bimboterminator1 replied:
Have totally removed this comment.

Stolb27 commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -1984,6 +1984,17 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+	 */
+	if (cte->cterefcount > 1 &&
+		((Query *) cte->ctequery)->commandType != CMD_SELECT)
+		elog(ERROR, "too much refs to non-SELECT CTE");
```
I think it's quite rough condition. As far I understand, the problem manifests itself only in case of inlining CTEs (see condition below).

bimboterminator1 replied:
Have moved this if-clause, what allows to generate plan when ` gp_cte_sharing` is on.

Stolb27 replied:
It was decided to separate these changes to a separate patch.

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1237,25 +1269,92 @@ cdbpath_motion_for_join(PlannerInfo *root,
+				/*
+				 * For the case that other is a Hashed table and redistribute
+				 * clause matches other's distribute keys and , we could
+				 * redistribute Replicated to other.
```
Do you have a test case for this branch? How to redistribute Replicated with numsegments > 1?

bimboterminator1 replied:
```
Test case in with_clause.sql
with_dml_dr_seg1 - at 1 segment 
with_dml - hashed at 3 segments

explain (costs off)
with cte as (
    insert into with_dml_dr_seg1 select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join with_dml on cte.i = with_dml.i;
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice2; segments: 3)
         ->  Aggregate
               ->  Hash Join
                     Hash Cond: (with_dml.i = with_dml_dr_seg1.i)
                     ->  Seq Scan on with_dml
                     ->  Hash
                           ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                 Hash Key: with_dml_dr_seg1.i
                                 ->  Insert on with_dml_dr_seg1
                                       ->  Function Scan on generate_series i
 Optimizer: Postgres query optimizer
(12 rows)
```
For case Replicated with numsegments > 1 (still has to be less than Hashed's numsegments to be redistributed):
```
explain (costs off)
with cte as (
    insert into t_repl_s2 select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join t_hashed on cte.i = t_hashed.i;
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice2; segments: 3)
         ->  Aggregate
               ->  Hash Join
                     Hash Cond: (t_hashed.i = t_repl_s2.i)
                     ->  Seq Scan on t_hashed
                     ->  Hash
                           ->  Redistribute Motion 2:3  (slice1; segments: 2)
                                 Hash Key: t_repl_s2.i
                                 ->  Insert on t_repl_s2
                                       ->  Function Scan on generate_series i
 Optimizer: Postgres query optimizer
(12 rows)
```

Stolb27 replied:
How does `Redistribute Motion 2:3` handle duplicate rows on input?

bimboterminator1 replied:
Probably one need to generate `Redistirbutre Motion 1:3` in this situation

Stolb27 replied:
But you must execute the Insert node on both segments. And they both will send the same tuples. I think that redistribution of replicated data sets must be excepted. The only that we can do and focus them or join them as is when possible.

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1109,31 +1140,32 @@ cdbpath_motion_for_join(PlannerInfo *root,
+			                    CdbPathLocus_NumSegments(segGeneral->locus);
+
+			    }
 				return other->locus;
```
In the end, I think you mustn't blindly use the number of segments Replicated locus especially in case of Segment General distributed on less number of segments:

```
i2160=# explain with cte1 as (update d2 set c1 = 15 returning *) select * from cte1 join d1 using(c1);
                                      QUERY PLAN
--------------------------------------------------------------------------------------
 Explicit Gather Motion 2:1  (slice1; segments: 2)  (cost=0.03..1.07 rows=4 width=28)
   ->  Hash Join  (cost=0.03..1.07 rows=2 width=28)
         Hash Cond: (d1.c1 = d2.c1)
         ->  Seq Scan on d1  (cost=0.00..1.00 rows=1 width=16)
         ->  Hash  (cost=1.01..1.01 rows=1 width=22)
               ->  Update on d2  (cost=0.00..1.00 rows=1 width=22)
                     ->  Seq Scan on d2  (cost=0.00..1.00 rows=1 width=22)
```

bimboterminator1 replied:
```suggestion
                               other->locus.numsegments =
			            CdbPathLocus_CommonSegments(segGeneral->locus,
                                                                other->locus);
				return other->locus;
```
(and remove above if-clause on numsegments)
Will it be allright? 
```
explain (costs off)
with cte as (
    insert into t_repl select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join t_repl_s2 on cte.i = t_repl_s2.i;
                            QUERY PLAN                            
------------------------------------------------------------------
 Explicit Gather Motion 2:1  (slice1; segments: 2)
   ->  Aggregate
         ->  Hash Join
               Hash Cond: (t_repl_s2.i = t_repl.i)
               ->  Seq Scan on t_repl_s2
               ->  Hash
                     ->  Insert on t_repl
                           ->  Function Scan on generate_series i
 Optimizer: Postgres query optimizer
(9 rows)


explain (costs off)
with cte as (
    insert into t_repl_s2 select i, i * 100 from generate_series(1,5) i
    returning i
) select count(*) from cte join t_repl on cte.i = t_repl.i;
                            QUERY PLAN                            
------------------------------------------------------------------
 Explicit Gather Motion 2:1  (slice1; segments: 2)
   ->  Aggregate
         ->  Hash Join
               Hash Cond: (t_repl.i = t_repl_s2.i)
               ->  Seq Scan on t_repl
               ->  Hash
                     ->  Insert on t_repl_s2
                           ->  Function Scan on generate_series i
 Optimizer: Postgres query optimizer
(9 rows)


```

Stolb27 replied:
> Will it be allright?

As far I can imagine, nope. In my example, r1 table is distributed on the only segment (less than d2 in the common case). Your changes make numsegments to the one (as for r1), but slice will be executed on number of segments for d2. Which does segment return rows?

bimboterminator1 replied:
The only way is to broadcast SegmentGeneral or make both SingleQE

Stolb27 commented:
What the Bolsheviks have been talking about for a long time. Volatile function violates Replicated distribution because different segments produce different output. This patch should include correct handling of volatile functions (see cc35273 for details):

```
i2160=# explain with cte1 as (update d2 set c1 = 15 returning *) select * from cte1 join r2 using(c1) where cte1.c2 > 10 * random();
                                           QUERY PLAN
-------------------------------------------------------------------------------------------------
 Gather Motion 2:1  (slice1; segments: 2)  (cost=0.04..1.08 rows=4 width=28)
   ->  Hash Join  (cost=0.04..1.08 rows=2 width=28)
         Hash Cond: (r2.c1 = cte1.c1)
         ->  Seq Scan on r2  (cost=0.00..1.00 rows=1 width=16)
         ->  Hash  (cost=1.01..1.01 rows=1 width=22)
               ->  Subquery Scan on cte1  (cost=0.00..1.01 rows=1 width=22)
                     Filter: ((cte1.c2)::double precision > ('10'::double precision * random()))
                     ->  Update on d2  (cost=0.00..1.00 rows=1 width=22)
                           ->  Seq Scan on d2  (cost=0.00..1.00 rows=1 width=22)
```

bimboterminator1 committed:
Allow gathering FLOW_REPLICATED for SELECT stmt

These changes allow execute SELECT query from CTE containing writable
DML on replicated table.

bimboterminator1 committed:
Allow gathering Replicated locus

These changes enables Gather Motion creation on the top of locus
Replicated. This is required for most of the joins, there locus
replicated is moved to the SingleQE. It also affects the ORDER BY
clause, changing the assert inside make_motion_gather enables the
sorting for Replicated results.

bimboterminator1 committed:
Join Replicated and SegmentGeneral

These changes add the support of join between result of writable
CTE query over replicated table and other replicated table. Both
relations are joined at SingleQE.

bimboterminator1 committed:
Support of joins with Replicated locus

These changes introduce the logic, which performs correct joins with
locus Replicated.
Entry/SingleQE + Replicated -> Entry/SingleQE
General + Replicated -> Replicated
Hashed/Strewn + Replicated if numsegments1 = numsegments2 && ! outer_join ->
Hashed/Strewn
Hashed/Strewn + Replicted if numsegs1 != numsegs2 || outer_join -> SingleQE

bimboterminator1 commented:
For convinient review the changes have been split into 6 commits.

The final patch description is inside the PR desc.

The diffs should be viewed in GitHub with **Hide whitespace** option

bimboterminator1 committed:
Volatile quals handling

bimboterminator1 committed:
Avoid broadcast of Replicated locus inside the SubPlans

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/55609

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/55662

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/55666

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/671027

InnerLife0 commented on src/backend/cdb/cdbllize.c:
```diff
@@ -590,8 +590,26 @@ ParallelizeCorrelatedSubPlanMutator(Node *node, ParallelizeCorrelatedPlanWalkerC
+				contain_volatile_functions((Node *) scanPlan->qual))
+			{
+				focusPlan(scanPlan, false /* stable */ , false /* rescannable */ );
+				Plan	   *res = (Plan *) make_result((PlannerInfo *) ctx->base.node, resTL, NULL, scanPlan);
```
Should move declaration upper to respect code agreements.

bimboterminator1 replied:
I've changed approach in https://github.com/arenadata/gpdb/commit/8b65137bdd5639df8afe1023d78f72786af0cb7c and https://github.com/arenadata/gpdb/commit/6ff803b990f7d5b1e56ad9996f0e6b2257aa3d27

InnerLife0 commented on src/test/regress/expected/with_clause_optimizer.out:
```diff
@@ -2847,3 +2847,80 @@ with cte as (
 drop table t_hashed_seg2;
+-- Test proper handling of the volatile conditions, which are
+-- applied to the result of modifying CTEs over replicated tables.
+-- Before the volatile filtration the result should be gathered.
```
I think this comment should be reworked as we filter before gathering.

InnerLife0 commented on src/test/regress/expected/with_clause_optimizer.out:
```diff
@@ -2847,3 +2847,80 @@ with cte as (
+    insert into with_dml_dr
+    select i, i * 100 from generate_series(2,5) i
+    returning i
+) select count(*) from cte where cte.i > 1 * random();
```
No need to multiple a result of random() to 1 :-)

InnerLife0 commented on src/test/regress/expected/with_clause.out:
```diff
@@ -3,6 +3,9 @@
 -- m/ERROR:  Too much references to non-SELECT CTE \(allpaths\.c:\d+\)/
 -- s/\d+/XXX/g
 --
+-- m/ERROR:  Cannot broadcast Replicated locus \(cdbllize\.c:\d+\)/
```
Where are the tests?

InnerLife0 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1108,134 +1108,143 @@ cdbpath_motion_for_join(PlannerInfo *root,
 				 */
-				Assert(CdbPathLocus_NumSegments(segGeneral->locus) <=
-					   CdbPathLocus_NumSegments(other->locus));
+				if (root->upd_del_replicated_table > 0 &&
```
I miss a test case for this branch.
In a voice chat you said, that you was unable to find a proper example. I don't have a firm decision about should we keep this branch or not. @Stolb27 What do you think?

bimboterminator1 replied:
I've removed that in 98043f6739acd887b49a7fb38ec04e8d5c1cb66c

InnerLife0 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -1246,28 +1246,60 @@ cdbpath_motion_for_join(PlannerInfo *root,
+				 * bring locus Replicated to singleton in order to perform join
+				 * on single segment
+				 */
+				replicated->move_to = other->locus;
```
My suggestion is to use CdbPathLocus_MakeSimple() here to unify approaches.

bimboterminator1 replied:
Done in 98043f6739acd887b49a7fb38ec04e8d5c1cb66c

InnerLife0 commented on src/backend/cdb/cdbllize.c:
```diff
@@ -590,8 +590,26 @@ ParallelizeCorrelatedSubPlanMutator(Node *node, ParallelizeCorrelatedPlanWalkerC
+			 * Therefore, in case of volatile quals we must focus Replicated
+			 * plan first, apply the quals, and then broadcast it back.
+			 */
+			if (scanPlan->flow->locustype == CdbLocusType_Replicated &&
```
During the private talk with @Stolb27 , I figured out that promising solution was to throw an error here with possible creation of explicit broadcast in the future.
Current solution works, but looks like a good old crutch. Why you switched?

bimboterminator1 replied:
I've changed approach in 8b65137bdd5639df8afe1023d78f72786af0cb7c and 6ff803b990f7d5b1e56ad9996f0e6b2257aa3d27

InnerLife0 commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -3,6 +3,9 @@
 --
+-- m/ERROR:  Cannot broadcast Replicated locus \(cdbllize\.c:\d+\)/
+-- s/\d+/XXX/g
+--
```
I think we miss some tests here or the result of existed tests should be changed.

bimboterminator1 replied:
Changed the tests

InnerLife0 commented:
Looks OK in general.
What I still want to see here:
1. As we talk about commit-by-commit review, PR description points count and commits count should be equal.
2. Commits description is poor now. You may push an empty commit with a description, which will be used on squash later.
________

A few words about what I see.
Historically GP is an MPP database. So, near the always, the best scenario for GP was to utilize all segments, correctly distribute data on all segments and do less on one segment.
With the advent of replicated tables, the count of motions became less, but scenarios have expanded with one segment only special cases.
The case with SELECT over DML kind of tears apart both of them, as we need to do DML over replicated table on all segments, but SELECT only on one.
So, I see that current solution is a compromise. Sometimes we do a double job, trying to filter or calc something on all segments. I'm OK with that and I think that currently we can't do more without much of changes.
___________
We already said about explicit broadcast, but did we take a look to one time filters?
Sometimes ORCA use makeSegmentFilterExpr() function to filter execution segments. Can we implement something like that for postgres planner?
```
explain (costs off, verbose, analyze)
with cte as (select i from with_dml_dr)
select count(*) from cte join
(select a from generate_series(1,5) a) x on cte.i = x.a;
                                                  QUERY PLAN                                                   
---------------------------------------------------------------------------------------------------------------
 Aggregate (actual time=6.116..6.116 rows=1 loops=1)
   Output: count()
   ->  Gather Motion 3:1  (slice1; segments: 3) (actual time=3.281..6.101 rows=18 loops=1)
         ->  Hash Join (actual time=1.549..2.595 rows=18 loops=1)
               Hash Cond: (generate_series.generate_series = with_dml_dr.i)
               Executor Memory: 2kB  Segments: 3  Max: 1kB (segment 0)
               work_mem: 2kB  Segments: 3  Max: 1kB (segment 0)  Workfile: (0 spilling)
               Extra Text: (seg0)   Hash chain length 3.6 avg, 4 max, using 5 of 262144 buckets.
               ->  Result (actual time=0.047..0.048 rows=5 loops=1)
                     Output: generate_series.generate_series
                     One-Time Filter: (gp_execution_segment() = 0)
                     ->  Function Scan on pg_catalog.generate_series (actual time=0.038..0.039 rows=5 loops=1)
                           Output: generate_series.generate_series
                           Function Call: generate_series(1, 5)
                           work_mem: 17kB  Segments: 1  Max: 17kB (segment 0)
               ->  Hash (actual time=0.026..0.026 rows=18 loops=1)
                     Output: with_dml_dr.i
                     ->  Seq Scan on public.with_dml_dr (actual time=0.008..0.012 rows=18 loops=1)
                           Output: with_dml_dr.i
 Planning time: 37.195 ms
   (slice0)    Executor memory: 127K bytes.
   (slice1)    Executor memory: 2127K bytes avg x 3 workers, 2139K bytes max (seg0).  Work_mem: 17K bytes max.
 Memory used:  128000kB
 Optimizer: Pivotal Optimizer (GPORCA)
 Execution time: 7.129 ms
(25 rows)
```

bimboterminator1 committed:
Partial changes to join logic

1. Join Replicated + SegmentGeneral
If Replicated's number of segments is less or equal than SegmentGeneral's,
we allow to perform join locally at Replicated's number of segments
(SegmentGenerals's number of segments can be safely reduced). Otherwise,
join is held at SingleQE. The commit has added appropriate tests.

2. Added the tests to Replicated+Replicated case. First of all, the
main assumption here is that it's impossible to have Replicated operands
be propagated at different number of segments. Replicated locus
can originally occur from 1)CTE with DML on replicated table; 2) Join of
Replicated locus with SegmentGeneral. Therefore in both cases, the
number of segments of Replicated locus can't be reduced.

bimboterminator1 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-2160-2

bimboterminator1 committed:
build fix

bimboterminator1 committed:
Revert volatility tests

bimboterminator1 committed:
Forbidden volatile functions over Replicated locus for joins

This commit introduces the variant of volatile functions handling,
when Replicated locus is joined with others. In order to avoid the
Explicit Gather - Broadcast/Redistribute pair, where 2 motions in
the plan are one on the top of the other, we decide to throw an
error in cases, when any of join operands has Replicated locus
and the path contains volatile functions inside.
path_contains_volatile_functions function was added to pathnode.c,
this function tries to cover vast variety of queries and checks
whether there are volatile functions in the path's restrictinfo,join
restrictinfo, targetlist and qual of subqueries' subplans.

bimboterminator1 committed:
Volatile functions handling for correlated/uncorrelated Sublans

1. The volatile functions were forbidden for Subplans. The logic
is implemented the same way as it's done for joins. For correlated
SubPlans the conditions are duplicating code from previous commit,
because the ParallelizeCorrelatedSubPlanMutator walks plan himself.
For uncorrelated Subplans the plan_contains_volatile_functions is
called.

2. For the case when Replicated locus is at different number of segments
 from the main plan, the same error is also thrown.

bimboterminator1 committed:
Support of UNION ALL Replicated+SegmentGeneral

bimboterminator1 committed:
fix comments

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60006

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864843

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864844

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864841

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864836

bimboterminator1 committed:
fix test run

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864971

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864972

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864964

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864963

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/864965

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/865561

bimboterminator1 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-2160-2

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60033

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/866434

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/866435

bimboterminator1 committed:
fix formatting

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/868105

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/868106

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/868099

InnerLife0 commented:
As it was decided on the voice chat, we want to implement a solution which will handle all replicated+volatile cases at once. This solution should be much simpler than trying to improve all possible error logic. Though, we may improve solution later if there will be any need.

bimboterminator1 committed:
Revert "Volatile functions handling for correlated/uncorrelated Sublans"

bimboterminator1 committed:
Revert "Forbidden volatile functions over Replicated locus for joins"

This reverts commit 8b65137bdd5639df8afe1023d78f72786af0cb7c.

bimboterminator1 committed:
Prohibit volatile functions to locus Replicated

bimboterminator1 committed:
Broadcast control for CTEs inside SubPlans

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60780

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/909223

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/909224

bimboterminator1 committed:
Extend UNION ALL case. additional fix and tests

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60783

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/909320

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/909321

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/909314

bimboterminator1 committed:
fix comment

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60789

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/909749

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/909750

InnerLife0 commented on src/backend/cdb/cdbllize.c:
```diff
@@ -758,8 +766,17 @@ ParallelizeSubplan(SubPlan *spExpr, PlanProfile *context)
+			/*
+			 * Broadcasting Replicated locus leads to data duplicates.
+			 */
+			if (newPlan->flow->locustype == CdbLocusType_Replicated &&
```
Can this be moved out of if statement (with and the same block from mutator)? If no, what non-error plans we may cut with this condition?

bimboterminator1 replied:
`conainingPlanDistributed` [indicates](https://github.com/arenadata/gpdb/blob/38c8f80aad843ef11966e3b44f17d2fc5b528363/src/backend/cdb/cdbllize.c#L790) whether the plan needs to be broadcasted.
When it's false the plan is focused (here doesn't matter how much segments at locus Replicated):
```
create table t_repl (i int, j int) distributed replicated;
create table t_repl2 (i int, j int) distributed replicated;

 explain (costs off) with cte as (insert into t_repl values (1,1) returning *)
 select * from t_repl2 where t_repl2.i in (select i from cte);
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Gather Motion 1:1  (slice2; segments: 1)
   ->  Seq Scan on t_repl2
         Filter: (hashed SubPlan 1)
         SubPlan 1  (slice2; segments: 1)
           ->  Materialize
                 ->  Explicit Gather Motion 3:1  (slice1; segments: 3)
                       ->  Subquery Scan on cte
                             ->  Insert on t_repl
                                   ->  Result
 Optimizer: Postgres query optimizer
(10 rows)
```

InnerLife0 replied:
I meant the something like:
```
if (containingPlanDistributed && newPlan->flow->locustype == CdbLocusType_Replicated &&
		newPlan->flow->numsegments != context->currentPlanFlow->numsegments)
			elog(ERROR, "could not parallelize SubPlan");
```
before or after the if statement I mentioned.

InnerLife0 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -2091,6 +2124,13 @@ turn_volatile_seggen_to_singleqe(PlannerInfo *root, Path *path, Node *node)
+		 * volatile functions handling.
+		 */
+		if (CdbPathLocus_IsReplicated(path->locus))
+			elog(ERROR, "could not devise a plan");
```
I think it's worth to move it down and use else-if here. The comment may be moved inside statement.

bimboterminator1 replied:
Done

InnerLife0 commented on src/backend/optimizer/plan/planner.c:
```diff
@@ -941,6 +950,15 @@ subquery_planner(PlannerGlobal *glob, Query *parse,
+	 * Replicated locus is not supported yet in context of
+	 * volatile functions handling.
+	 */
+	if (plan->flow->locustype == CdbLocusType_Replicated &&
```
The same suggestion about else-if.

bimboterminator1 replied:
Done

bimboterminator1 committed:
Move elog under else if block

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60997

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/923196

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/923197

bimboterminator1 committed:
Generate more data for joins

bimboterminator1 committed:
Insert data for SubPlan and Union  tests

bimboterminator1 committed:
Fix test output

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61172

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932280

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932281

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/932272

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-2160-2

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/61250

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/936564

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/936565
